lib/ws/ws.d.ts

```diff
@@ -0,0 +1,258 @@
+import {EventEmitter} from "events";
+
+import { BadMessageError, NoMethodHandlerError } from "../errors";
+
+declare class BeamSocket extends EventEmitter {
+  /**
+   * Constructor for the class to create the socket handle / connection.
+   */
+  constructor(addresses: string[]);
+  /**
+   * Which connection we use in our load balancing.
+   */
+  private _addressOffset: number;
+  /**
+   * List of addresses we can connect to.
+   */
+  private _addresses: string[];
+  /**
+   * Spool to store events queued when the connection is lost.
+   */
+  private _spool: string[];
+  /**
+   * The WebSocket instance we're currently connected with.
+   */
+  ws: WebSocket;
+  /**
+   * The status of the socket connection.
+   */
+  status: ISocketStatus;
+  /**
+   * Counter of the current number of reconnect retries, and the number of retries before we reset our reconnect attempts.
+   */
+  private _retries: number;
+  /**
+   * Max 2 minute retry time.
+   */
+  private _retryWrap: number;
+  /**
+   * Timeout waiting to reconnect
+   */
+  private _reconnectTimeout: any;
+  /**
+   * Number of milliseconds to wait in .call() before we give up waiting or the reply. Used to prevent leakages.
+   */
+  callTimeout: number;
+  /**
+   * Map of call IDs to promises that should be resolved on method responses.
+   */
+  private _replies: {};
+  /**
+   * Authentication packet store that we'll resend if we have to reconnect.
+   */
+  private _authpacket: string;
+  /**
+   * Counter for method calls.
+   */
+  private _callNo: number;
+  /**
+   * The timeout error for the socket.
+   */
+  TimeoutError: TimeoutError;
+  /**
+   * Set the socket status values.
+   */
+  IDLE: number;
+  CONNECTED: number;
+  CLOSING: number;
+  CLOSED: number;
+  CONNECTING: number;
+  /**
+   * Gets the status of the socket connection.
+   */
+  getStatus(): ISocketStatus;
```

Comment:
I would change this to just `number` and remove the `ISocketStatus` enum as this is not actually an enum.